### ISSUE TITLE
gRPC gateway question

### ISSUE THREAD

**Issue by EwanValentine** [micro/examples#13]
Me again, 

I'm comparing using gRPC gateway, vs the preferred go-micro web-rpc approach, and why you should just use the web-rpc approach. But I'm having trouble explaining something. 

I noticed when using grpc-gateway, you have to do `grpc.NewService()` such as in the examples, and `micro.NewService()` doesn't work with grpc-gateway. But if you use `grpc.NewService()`, that becomes incompatible with services which use `micro.NewService()`. 

I think you briefly touched on this in a previous ticket I raised. But am I right in assuming that you lose much of go-micro's functionality when using the `micro/go-grpc` library because it has to wrap around standard grpc, rather than the go-micro grpc stuff? 

Cheers!

**Comment by asim** [comment]
I'm not sure what you mean by web-rpc. There's nothing called web-rpc in the tooling. I'm going to assume you mean the micro api. The micro api is an api gateway that provides dynamic routing to any service based on the path of the request. The reason we use this over something like a generated gateway is because a microservice environment is dynamic, we want to route to new services based on service discovery, not have to rebuild a generated gateway for new services or new request formats. 

On the grpc-gateway itself. We were getting a lot of requests for integration. What people didn't understand is that grpc itself is a protocol and the client/server plugins are fully compatible with the grpc protocol. So integrating the grpc-gateway is as simple as generating it and pointing it towards a service. We made it easier to write a service that uses grpc with the go-grpc library. If you look at the code, it's literally less than a 30 line file that just initialises the client/server plugins.

Here's the whole thing

```
package grpc

import (
	"time"

	"github.com/micro/go-micro"

	client "github.com/micro/go-plugins/client/grpc"
	server "github.com/micro/go-plugins/server/grpc"
)

// NewService returns a grpc service compatible with go-micro.Service
func NewService(opts ...micro.Option) micro.Service {
	// our grpc client
	c := client.NewClient()
	// our grpc server
	s := server.NewServer()

	// create options with priority for our opts
	options := []micro.Option{
		micro.Client(c),
		micro.Server(s),
	}

	// append passed in opts
	options = append(options, opts...)

	// generate and return a service
	return micro.NewService(options...)
}
```

Here's what to remember, go-micro is a pluggable RPC framework. It comes with defaults such as a http transport and json-rpc or proto-rpc for message encoding. These are plugins we can swap out. When you do `micro.NewService()` you're getting these defaults. You can't just expect this to automatically work with grpc. You either have to swap in the grpc plugins or use the convenient library that just lets you do `grpc.NewService()`. You don't lose go-micro functionality when using go-grpc. Look at the return type of the `NewService` function. It returns a `micro.Service`. It takes `micro.Option` as its arguments. It's fully compatible. If you read the readme you'll see it says you only need to change one line, everything else works the same.

Hope that helps.

**Comment by EwanValentine** [comment]
Apologies, I was using the wrong name, I did mean micro api. 

Yes that all makes sense I think. And that does confirm what I initially thought. My confusion around this stemmed from an issue I had whereby I had a cli program which worked with a program instantiated via `micro.NewService()`, but when I changed it to `grpc.NewService()` in order to interface grpc-gateway, I would get an error on the client:
```javascript
{"id":"go.micro.client.transport","code":500,"detail":"malformed HTTP response \"\\x00\\x00\\x00\\x04\\x00\\x00\\x00\\x00\\x00\"","status":"Internal Server Error"}
```

So in order for a go-micro cmd client to function with this in parallel with a grpc-gateway api, do I need to change anything in the way the client's ran/configured, based on what you've just said around defaults etc? It's probably something really obvious, but I just wanted to triple check I understood.

Thanks for your time and patience!

**Comment by asim** [comment]
TLDR; `go get github.com/micro/grpc/cmd/micro`

Ok so rather than just telling you what to do, I want you to be able to make sense of this so it carries forward into other stuff you use.

The micro api, cli and all other components of the toolkit which call a service make use of the client. The client is used to make requests, a service is used to serve requests. If a service is using the grpc server plugin, it means its using grpc for its communication mechanism. To speak to this server we need the equivalent grpc client plugin. When you initialise a service using go-grpc it's initialising both the server and client plugins so that it can handle requests via grpc but also make them to other services via grpc. 

So as mentioned before, go-micro has default settings of http transport, json-rpc and proto-rpc for encoding. The micro toolkit makes use of the go-micro client to make any requests which means in this case its using the default client and that's why you're seeing that error. In order to speak to services using grpc you must use the grpc plugin within the micro toolkit. It's the same for any other plugins for service discovery, encoding, etc, etc. 

If you look at the go-grpc readme you'll see an explanation of how to include that with the micro toolkit https://github.com/micro/go-grpc#using-with-micro-toolkit

There's also a repo that presets this for you https://github.com/micro/grpc where you can just do `go get github.com/micro/grpc/cmd/micro`

Further instructions which I've taken from the readme just so you don't have to do anything

## Using with Micro toolkit

You may want to use the micro toolkit with grpc services. To do this either use the prebuilt toolkit or 
simply include the grpc client plugin and rebuild the toolkit.

### Go Get

```
go get github.com/micro/grpc/cmd/micro
```

### Build Yourself

```
go get github.com/micro/micro
```

Create a plugins.go file
```go
package main

import _ "github.com/micro/go-plugins/client/grpc"
import _ "github.com/micro/go-plugins/server/grpc"
```

Build binary
```shell
// For local use
go build -i -o micro ./main.go ./plugins.go
```

Flag usage of plugins
```shell
micro --client=grpc --server=grpc
```

**Comment by EwanValentine** [comment]
Gotcha! This makes perfect sense now, thanks for taking the time to explain that more. Appreciate it's already documented in places, apologies for missing that. I'll do some more reading.

Cheers!

**Comment by lpy-neo** [comment]
I also run into this problem, and it makes me crazy. I think it is much better to give a hint about the client and server options in the document of the greeter example.
